plugins/processors/lotus_peer_id/peerid.go

```diff
@@ -0,0 +1,163 @@
+
+type Exclude struct {
+	Measurement string `toml:"measurement"`
+}
+
+type LotusPeerIDTagger struct {
+	Config_DataPath      string `toml:"lotus_datapath"`
+	Config_APIListenAddr string `toml:"lotus_api_multiaddr"`
+	Config_APIToken      string `toml:"lotus_api_token"`
+
+	apiBackOffThrottle time.Duration
+	curAPIBackOff time.Time
+
+	initializedDefaults bool
+
+	Excludes []Exclude `toml:"exclude"`
+
+	excludeMap map[string]struct{}
+
+	peerID peer.ID
+}
+
+func (l *LotusPeerIDTagger) setAPIConfigDefaults() {
+	if len(l.Config_DataPath) == 0 {
+		l.Config_DataPath = DefaultConfig_DataPath
+	}
+	if len(l.Config_APIListenAddr) == 0 {
+		l.Config_APIListenAddr = DefaultConfig_APIListenAddr
+	}
+	l.apiBackOffThrottle = time.Second * 10
+	l.curAPIBackOff = time.Unix(0,0)
+}
+
+func (l *LotusPeerIDTagger) setPeerID() error {
+	nodeAPI, nodeCloser, err := l.getAPIUsingLotusConfig()
+	if err != nil {
+		return err
+	}
+	defer nodeCloser()
+	pid, err := nodeAPI.ID(context.Background())
+	if err != nil {
+		return err
+	}
+	l.peerID = pid
+	return nil
+}
+
+func (l *LotusPeerIDTagger) setExcludeMap() {
+	ex := make(map[string]struct{})
+	for _, e := range l.Excludes {
+		ex[e.Measurement] = struct{}{}
+	}
+	l.excludeMap = ex
+}
+
+func (l *LotusPeerIDTagger) getAPIUsingLotusConfig() (api.FullNode, func(), error) {
+	var (
+		nodeAPI    api.FullNode
+		nodeCloser func()
+	)
+	if len(l.Config_APIListenAddr) > 0 && len(l.Config_APIToken) > 0 {
+		api, apiCloser, err := rpc.GetFullNodeAPIUsingCredentials(l.Config_APIListenAddr, l.Config_APIToken)
+		if err != nil {
+			err = fmt.Errorf("connect with credentials: %v", err)
+			return nil, nil, err
+		}
+		nodeAPI = api
+		nodeCloser = apiCloser
+	} else {
+		api, apiCloser, err := rpc.GetFullNodeAPI(l.Config_DataPath)
+		if err != nil {
+			err = fmt.Errorf("connect from lotus state: %v", err)
+			return nil, nil, err
+		}
+		nodeAPI = api
+		nodeCloser = apiCloser
+	}
+	return nodeAPI, nodeCloser, nil
+}
+
+func (l *LotusPeerIDTagger) SampleConfig() string {
+	return sampleConfig
+}
+
+func (l *LotusPeerIDTagger) Description() string {
+	return "Decorate measurements that pass through this filter with the peer id of lotus daemon."
+}
+
+func (l *LotusPeerIDTagger) Apply(in ...telegraf.Metric) []telegraf.Metric {
+	if !l.initializedDefaults {
+		// configure lotus connection
+		l.setAPIConfigDefaults()
+		// create mapping of excluded metrics for faster processing.
+		l.setExcludeMap()
+		l.initializedDefaults = true
+	}
+
+	if time.Since(l.curAPIBackOff) < l.apiBackOffThrottle{
+		return nil
+	}
+
+	if l.peerID.Size() == 0 {
+		log.Println("Setting up lotus_peer_id processor")
+		// extract peerID from connected lotus daemon
+		if err := l.setPeerID(); err != nil {
+			log.Println("!E Failed to get peerID from lotus API", err.Error())
+			l.curAPIBackOff = time.Now()
+			log.Println("!E Back off for", l.apiBackOffThrottle.String())
+		}
+		return nil
+	}
```

Comment:
Note: This logic means we will drop metrics if we cannot reach the lotus daemon

Reply:
> requirement above
Referring to https://github.com/filecoin-shipyard/telegraf-lotus/pull/13#discussion_r467240733